content/blog/2018-03-27-update-on-async-rendering.md

```diff
@@ -0,0 +1,189 @@
+`embed:update-on-async-rendering/updating-external-data-when-props-change-after.js`
+
+> Note
+>
+> If you're using an HTTP library that supports cancellation, like [axios](https://www.npmjs.com/package/axios), then it's simple to cancel an in-progress request when unmounting. For native Promises, you can use an approach like [the one shown here](https://gist.github.com/bvaughn/982ab689a41097237f6e9860db7ca8d6).
+
+### Reading DOM properties before an update
+
+Here is an example of a component that reads a property from the DOM before an update in order to maintain scroll position within a list:
+`embed:update-on-async-rendering/react-dom-properties-before-update-before.js`
+
+In the above example, `componentWillUpdate` is used to read the DOM property. However with async rendering, there may be delays between "render" phase lifecycles (like `componentWillUpdate` and `render`) and "commit" phase lifecycles (like `componentDidUpdate`). If the user does something like resize the window during this time, the `scrollHeight` value read from `componentWillUpdate` will be stale.
+
+The solution to this problem is to use the new "commit" phase lifecycle, `getSnapshotBeforeUpdate`. This method gets called _immediately before_ mutations are made (e.g. before the DOM is updated). It can return a value for React to pass as a parameter to `componentDidUpdate`, which gets called _immediately after_ mutations.
+
+The two lifecycles can be used together like this:
+
+`embed:update-on-async-rendering/react-dom-properties-before-update-after.js`
+
+## Other scenarios
+
+While we tried to cover the most common use cases in this post, we recognize that we might have missed some of them. If you are using `componentWillMount`, `componentWillUpdate`, or `componentWillReceiveProps` in ways that aren't covered by this blog post, and aren't sure how to migrate off these legacy lifecycles, please [file a new issue against our documentation](https://github.com/reactjs/reactjs.org/issues/new) with your code examples and as much background information as you can provide. We will update this document with new alternative patterns as they come up.
+
+## Open source project maintainers
+
+Open source maintainers might be wondering what these changes mean for shared components. If you implement the above suggestions, what happens with components that depend on the new static `getDerivedStateFromProps` lifecycle? Do you also have to release a new major version and drop compatibility for React 16.2 and older?
+
+Fortunately, you do not!
+
+In support of version 16.3, we've also created a new NPM package, [`react-lifecycles-compat`](https://github.com/reactjs/react-lifecycles-compat). This package polyfills components so that the new `getDerivedStateFromProps` lifecycle will also work with older versions of React (0.14.9+).
```

Comment:
Maybe: "When React 16.3 is published, we'll also publish a new npm package," since people aren't meant to use it yet?

(also npm should be lowercase https://css-tricks.com/start-sentence-npm/)

Reply:
Oh. I didn't know that (NPM vs npm)